### ISSUE TITLE
Post-quantum crypto?

### ISSUE THREAD

**Issue by tomato42** [latchset/tang#36]
My understanding of the protocol is that it currently depends on classical cryptography: FFC or ECC.

Most of algorithms in consideration in the [post-quantum selection process](https://csrc.nist.gov/Projects/Post-Quantum-Cryptography) do not operate on groups though. With exception of SIKE – Supersingular Isogeny Key Exchange (or SIDH).

Maybe we should consider implementing support for it? (using two algorithms in parallel, like in CECPQ1, may be a good idea)

**Comment by npmccallum** [comment]
I agree completely. I suspect any method which supports multi-party key exchange can be made to work.

**Comment by vicaya** [comment]
Tang protocol itself is stateless and works well independent of TLS. OTOH, it seems to me that with TLS, the standard ECDH exchange would suffice? i.e. eJWK stuff would be unnecessary?

Given TLS 1.3 now supports ML-KEM as the standard hybrid PQC, we could introduce a standard ECDH mode, which gets PQC for free over TLS?

**Comment by tomato42** [comment]
1. SIKE is broken so that shouldn't be used now
2. KEM and ML-KEM in particular are not key exchange algorithms, they are key encapsulation mechanisms, so they are not drop-in replacements for ECDH: e.g. you can't have a client publish its public ML-KEM in a certificate, server publish its public ML-KEM in a certificate, and have them able to communicate and confirm their identity through such a key exchange, with ECDH it is possible.

**Comment by vicaya** [comment]
By ML-KEM in TLS 1.3, I of course meant X25519MLKEM768, which use ML-KEM to wrap ECDH for transport only.

It's already supported by curl 8.15+ (OpenSSL 3.5.1+). Since the transport is already PQ safe, the tang protocol over `curl --curves X25519MLKEM768 https://tang-server.domain/...` is already PQ safe as well.

My comment also questioned the necessity of the M-R exchange that use the extra eJWK, when transport is already protected by TLS. The threat model of "leaking" cJWK (client public key) and dJWK (result of c * S) in standard ECDH to the server doesn't really apply in NBDE/KMS use cases, where tang servers should have higher trust level than clients anyway, since we already assume no root compromise on tang servers.

**Comment by clauderobi** [comment]
The main issue with TLS (post quantum or not) is that in reality is it always hop by hop, where each hop sees the clear text content and where the hops are not 100% control by either ends of the "virtual" (or proxied) TLS. So TLS is not enough. But ML-KEM can be applied above TLS. 

But, regardless of M-R (or ECMR) or not, the fact that an adversary, even using ML-KEM, can legitimately gain access to the server's public key is a problem, with QC it can recover the server's private key. In other words, just protecting the transmission does not help.

I am looking for a PQ solution; but ATM I can not find any. I am open for discussion though...

**Comment by vicaya** [comment]
> The main issue with TLS (post quantum or not) is that in reality is it always hop by hop, where each hop sees the clear text content and where the hops are not 100% control by either ends of the "virtual" (or proxied) TLS. So TLS is not enough

This is incorrect, since SNI was introduced as an extension in 2003, i.e., intermediate gateways can just do stream forwarding based on SNI instead of having to decode the content. You can indeed ensure end-to-end encryption with mTLS instead of relying on public CAs.

**Comment by clauderobi** [comment]
You ignore, WAF, CDN, Reverse proxy, DPI that are all able to decrypt the
traffic and in many cases will always do to perform their duty. Some of
them will do on a sampling basis. But nonetheless  have the capability to
intercept.

Bottom line, TLS is very very very rarely end-to-end and almost always
traverses intermediate points where it is terminated, decrypted and then
embarked in a new tunnel with whatever security settings you don't know.

mTLS is a non-starter for most cases, particularly those involving users.

**Comment by npmccallum** [comment]
I think it would be eminently reasonable to implement an OPRF version of Clevis/Tang. This would sacrifice offline provisioning that ECMR provides, but it aligns the usage towards a more standard algorithm that maintains most of the MR properties. There are even emerging PQ-OPRF algorithms such as Leap. This would not require TLS.

**Comment by clauderobi** [comment]
ML-KEM also does not require TLS. Between ML-KEM and PQ-OPRF, wha would require more processing?

**Comment by npmccallum** [comment]
Processing isn't really the point. It is the properties of the exchange.

**Comment by clauderobi** [comment]
Most of the time, I agree. But there may be use cases where it can become a
problem. But the properties are definitely paramount.
